### PR TITLE
Issue #6 solution

### PR DESCRIPTION
**Pull request by ajaskier**

### CODE REVIEW

File: gui/dist_view/distviewcompute.cpp [hunk 1]
```diff
@@ -81,20 +81,18 @@ multi_img::Value Compute::curpos(
...
-			pixel[d] = b.means[d] / b.weight;
 			std::pair<int, int> &range = ranges[d];
 			range.first = std::min<int>(range.first, (int)(it->first)[d]);
 			range.second = std::max<int>(range.second, (int)(it->first)[d]);
 		}
-		// TODO: calculate colors for all pixels BEFORE this step with functor
-		color = multi_img::bgr(pixel, meta, maxval);
-		b.rgb = QColor(color[2]*255, color[1]*255, color[0]*255);
+		for (int i = 0; i<3; i++) {
+			b.colors[i] = b.colors[i] / b.weight;
```
Comment (ypnos) [inline gui/dist_view/distviewcompute.cpp]: changing the variable in the bin at this stage messes up incremental updates later one (e.g. when labeling single pixels in the bandview)!
why not set b.rgb = QColor(b.colors[2] / b.weight …  directly?

File: gui/dist_view/distviewcompute.h [hunk 2]
```diff
@@ -55,8 +59,13 @@ struct Bin {
 					   std::minus<multi_img::Value>());
 	}
 
+	inline void sub(const cv::Vec3f &colorpixel) {
+		colors -= colorpixel;
+	}
+
 	float weight;
 	std::vector<multi_img::Value> means;
+	cv::Vec3f colors;
```
Comment (ypnos) [inline gui/dist_view/distviewcompute.h]: why "colors" instead of "color"? it is only a single color

File: gui/dist_view/distviewcompute.h [hunk 3]
```diff
@@ -47,6 +47,10 @@ struct Bin {
 					   std::plus<multi_img::Value>());
 	}
 
+	inline void add(const cv::Vec3f &colorpixel) {
```
Comment (ypnos) [inline gui/dist_view/distviewcompute.h]: please implement as a method with two arguments (pixel and color) that calls the other method (only pixel). so we don't have two add() calls in accumulate. same for sub.

File: gui/dist_view/distviewcompute.h [hunk 4]
```diff
@@ -132,6 +141,8 @@ struct ViewportCtx {
 
 	/* metadata depending on display configuration */
 	int nbins;
+
+	bool coloringUpToDate;
```
Comment (ypnos) [inline gui/dist_view/distviewcompute.h]: please call coloringValid and rename "valid" to "metadataValid", and group them together.

File: gui/dist_view/distviewmodel.cpp [hunk 5]
```diff
@@ -346,3 +345,24 @@ QPolygonF DistViewModel::getPixelOverlay(int y, int x)
...
+
+	if (coloringEnabled) {
+		updateBinning( (*context)->nbins );
+	}
+}
+
+void DistViewModel::processRgbToggled(bool enabled)
+{
+	coloringEnabled = enabled;
+	if( coloringEnabled && !(*context)->coloringUpToDate) {
```
Comment (ypnos) [inline gui/dist_view/distviewmodel.cpp]: this method accesses context without contextlock!

File: gui/dist_view/distviewmodel.cpp [hunk 6]
```diff
@@ -346,3 +345,24 @@ QPolygonF DistViewModel::getPixelOverlay(int y, int x)
...
+		updateBinning( (*context)->nbins );
```
Comment (ypnos) [inline gui/dist_view/distviewmodel.cpp]: spaces..

File: gui/dist_view/distviewmodel.h [hunk 7]
```diff
@@ -90,6 +94,9 @@ public slots:
 	 * (between subImage, addImage)
 	 */
 	bool inbetween;
+
+	bool coloringEnabled = false;
+	cv::Mat3f* coloringResult = nullptr;
```
Comment (ypnos) [inline gui/dist_view/distviewmodel.h]: call it coloring and use a cv::Mat3f, not pointer. you can test it with .empty(). cv::Mat types use memory sharing so there is no copy involved (only header) in passing them around. don't pass pointer here.

File: gui/model/falsecolor/falsecoloringcacheitem.h [hunk 8]
```diff
@@ -4,16 +4,19 @@
...
 	// Default copy constructor and assignment operator.
 
 	void invalidate() { pixmap_ = QPixmap(); }
 	bool valid() { return ! pixmap_.isNull(); }
 	QPixmap pixmap() { return pixmap_; }
+	cv::Mat3f* mat() { return mat_; }
+
 private:
 	QPixmap pixmap_;
+	cv::Mat3f* mat_;
```
Comment (ypnos) [inline gui/model/falsecolor/falsecoloringcacheitem.h]: same as in distviewmodel. don't use a pointer. also I don't see a delete anywhere!?

File: gui/model/falsecolor/falsecolormodelpayload.h [hunk 9]
```diff
@@ -57,6 +58,7 @@ private slots:
 	SharedMultiImgPtr grad;
 	CommandRunner *runner;
 	QPixmap result;
+	cv::Mat3f *matResult = nullptr;
```
Comment (ypnos) [inline gui/model/falsecolor/falsecolormodelpayload.h]: as discussed, don't use a pointer. please rename: resultMat and resultPixmap

File: gui/dist_view/distviewbinstbb.cpp [hunk 10]
```diff
@@ -17,9 +17,10 @@ class Accumulate {
 	Accumulate(bool subtract, multi_img &multi, const cv::Mat1s &labels, const cv::Mat1b &mask,
 		int nbins, multi_img::Value binsize, multi_img::Value minval, bool ignoreLabels,
 		std::vector<multi_img::Value> &illuminant,
-		std::vector<BinSet> &sets)
+		std::vector<BinSet> &sets, cv::Mat3f colormat)
```
Comment (ypnos) [inline gui/dist_view/distviewbinstbb.cpp]: please call the variable coloring or colors.

File: gui/dist_view/distviewbinstbb.cpp [hunk 11]
```diff
@@ -168,7 +171,12 @@ void Accumulate::operator()(const tbb::blocked_range2d<int> &r) const
 			if (subtract) {
 				BinSet::HashMap::accessor ac;
 				if (s.bins.find(ac, hashkey)) {
-					ac->second.sub(pixel);
+					if (!colormat.empty()) {
+						const cv::Vec3f &colorpixel = colormat(y,x);
```
Comment (ypnos) [inline gui/dist_view/distviewbinstbb.cpp]: just call it color or rgb here

File: gui/dist_view/distviewcompute.h [hunk 12]
```diff
@@ -30,7 +30,7 @@
  * hash key (the hash key is not part of the Bin class
  */
 struct Bin {
-	Bin() : weight(0.f) {}
+	Bin() : weight(0.f), color(0.f, 0.f, 0.f) {}
 	Bin(const multi_img::Pixel& initial_means)
 		: weight(1.f), means(initial_means) {} //, points(initial_means.size()) {}
```
Comment (ypnos) [inline gui/dist_view/distviewcompute.h:38]: the constructor takes initial means, I suppose it should also take an initial color mix. can you check where it is called?

Comment (ajaskier) [inline gui/dist_view/distviewcompute.h:38]: It's called nowhere

**Pull request closed without merging**